engine-api support

MoonShiesty opened the pull request:
This PR adds support to specular-geth to support deterministic execution of transactions via the engine-api.

To support this we need to make a few changes to geth:

- extend [`PayloadAttributes`](https://github.com/ethereum/execution-apis/blob/769c53c94c4e487337ad0edea9ee0dce49c79bfa/src/engine/specification.md#PayloadAttributesV1) with `Transactions`, a list of transaction to force include in the block
- extend [`PayloadAttributes`](https://github.com/ethereum/execution-apis/blob/769c53c94c4e487337ad0edea9ee0dce49c79bfa/src/engine/specification.md#PayloadAttributesV1) with `NoTxPool`,  if true no transactions are taken out of the tx-pool, only transactions from the above `Transactions` list will be included. to validate existing blocks we will exclude transactions from the mempool, to sequence new blocks we will include transactions from the mempool.
- allow proposers to reorg their own chain


Specular changes differ from optimism/superchain engine-api spec, because we don't use deposit/system transactions:

https://op-geth.optimism.io/

https://github.com/ethereum-optimism/optimism/blob/develop/specs/exec-engine.md#engine-api

siosw commented on beacon/engine/types.go:
```diff
@@ -32,15 +32,25 @@ import (
+	Transactions [][]byte `json:"transactions,omitempty"  gencodec:"optional"`
+	// NoTxPool is a field for rollups: if true, the no transactions are taken out of the tx-pool,
+	// only transactions from the above Transactions list will be included.
+	NoTxPool bool `json:"noTxPool,omitempty" gencodec:"optional"`
```
is there any place where the decision to not to include a gas-limit param is documented / explained?
i'm only aware of this https://github.com/ethereum/builder-specs/issues/29 which generally seems to favour including it

MoonShiesty replied:
Discussed on call this morning:

We're going to investigate if we should make this change, and what approach to take. But we can do it in a separate PR.

MoonShiesty commented:
@ujvl ready for review

I added an `L2EngineApi` param to the `ChainConfig`, which enables our L2 engine ap modifications. This change really helped reduce the specular diff.

Right now, `L2EngineApi` can be enabled/disabled with the genesis.json file. If it makes more sense to use a cli flag, it looks like we need to move this parameter to a different config.

I also cleaned up how we disable the ConsensusAPI heartbeat.

LEAFERx commented on miner/payload_building.go:
```diff
@@ -165,12 +179,19 @@ func (w *worker) buildPayload(args *BuildPayloadArgs) (*Payload, error) {
 	payload := newPayload(empty, args.Id())
+	// <specular modification>
+	if args.NoTxPool { // don't start the background payload updating job if there is no tx pool to pull from
+		return payload, nil
```
do you need to mark as full here? ref https://github.com/ethereum-optimism/op-geth/blob/optimism/miner/payload_building.go#L221

MoonShiesty replied:
fixed

LEAFERx commented on miner/worker.go:
```diff
@@ -880,9 +924,12 @@ func (w *worker) prepareWork(genParams *generateParams) (*environment, error) {
+	if len(w.extra) != 0 && w.chainConfig.L2EngineApi { // Specular chains must not set any extra data.
 		header.Extra = w.extra
 	}
+	// <specular modification>
```
missing `/`

MoonShiesty replied:
fixed

LEAFERx reviewed (commented):
1. What does Optimism's `RollupComputePendingBlock` option mean? Why don't we need it?
2. Can you explain several inconsistent changes w/ optimism:
    1. missing changes in https://github.com/ethereum-optimism/op-geth/blob/optimism/miner/worker.go#L559
    2. missing changes in https://github.com/ethereum-optimism/op-geth/blob/optimism/miner/worker.go#L1059
3. The signature of `getSealingBlock` is different. Is this caused by different geth versions?

ujvl commented on params/config.go:
```diff
@@ -304,6 +304,9 @@ type ChainConfig struct {
 	Clique    *CliqueConfig `json:"clique,omitempty"`
 	IsDevMode bool          `json:"isDev,omitempty"`
+	// <specular modification>
+	L2EngineApi bool `json:"l2EngineApi,omitempty"`
```
nit rename to
```suggestion
	EnableL2EngineAPI bool `json:"l2EngineApi,omitempty"`
```

ujvl commented on miner/worker_test.go:
```diff
@@ -468,7 +470,9 @@ func testGetSealingWork(t *testing.T, chainConfig *params.ChainConfig, engine co
 	for _, c := range cases {
-		block, _, err := w.getSealingBlock(c.parent, timestamp, c.coinbase, c.random, nil, false)
+		// <specular modification>
+		block, _, err := w.getSealingBlock(c.parent, timestamp, c.coinbase, c.random, nil, false, nil, nil)
```
might be misreading but is this setting a bool to nil (2nd to last param)

MoonShiesty replied:
i was actually passing an extra param. I did run `make test` to confirm all the tests pass. There must be a different way to build an run this test, I'll look into it.

MoonShiesty replied:
If we update to the latest geth, the new `getSealingBlock` api means we can eliminate the specular diff in this file.

ujvl commented on beacon/engine/types.go:
```diff
@@ -32,15 +32,25 @@ import (
+	Random                common.Hash    `json:"prevRandao"            gencodec:"required"`
+	SuggestedFeeRecipient common.Address `json:"suggestedFeeRecipient" gencodec:"required"`
+	// <specular modification>
+	Withdrawals []*types.Withdrawal `json:"withdrawals,omitempty" gencodec:"optional"`
```
i have the same q

MoonShiesty replied:
added this by mistake, fixed

MoonShiesty commented:
> 1. What does Optimism's `RollupComputePendingBlock` option mean? Why don't we need it?
> 2. Can you explain several inconsistent changes w/ optimism:
>    
>    1. missing changes in https://github.com/ethereum-optimism/op-geth/blob/optimism/miner/worker.go#L559
>    2. missing changes in https://github.com/ethereum-optimism/op-geth/blob/optimism/miner/worker.go#L1059
> 3. The signature of `getSealingBlock` is different. Is this caused by different geth versions?

1. `RollupComputePendingBlock` determines whether the miner computes a pending block: i.e. whether eth.getBlock('pending') returns the 'latest' block or a 'pending' block using the mempool. This shouldn't affect the engine-api payload/`engine_forkchoiceUpdatedV1` changes in the PR.
2. https://github.com/ethereum-optimism/op-geth/blob/optimism/miner/worker.go#L559 I didn't implement  `RollupComputePendingBlock` in this PR.
https://github.com/ethereum-optimism/op-geth/blob/optimism/miner/worker.go#L1059
Optimism passes a `GasLimit` as a part of the engine-api payload, I didn't implement the `GasLimit` parameter or make any changes to calculating a block's `GasLimit` as part of this PR.

3. op-geth updated to a later version of geth since I opened this PR. it might be easier to review these changes if I update to the latest geth.

LEAFERx commented on cmd/utils/flags.go:
```diff
@@ -930,6 +930,14 @@ Please note that --` + MetricsHTTPFlag.Name + ` must be set to start the server.
+	EnableL2EngineApiFlag = &cli.BoolFlag{
+		Name:     "enableL2EngineApi",
+		Usage:    "Enable L2 Engine API",
+		Category: flags.EthCategory,
```
eth is good for now
